src/io_posix.c

```diff
@@ -102,26 +102,32 @@ udx__recvmsg (udx_socket_t *handle, uv_buf_t *buf, struct sockaddr *addr, int ad
 void
 udx__on_writable (udx_socket_t *socket) {
 #ifdef UDX_PLATFORM_HAS_SENDMMSG
-  while (socket->send_queue.len > 0) {
+  udx_fifo_t *fifo = &socket->send_queue;
+
+  while (fifo->len > 0) {
     udx_packet_t *batch[UDX_SENDMMSG_BATCH_SIZE];
     struct mmsghdr h[UDX_SENDMMSG_BATCH_SIZE];
 
+    while (udx__fifo_peek(fifo) == NULL && fifo->len > 0) {
+      udx__fifo_shift(fifo);
+    }
+
+    if (fifo->len == 0) {
+      return;
+    }
+
     int pkts = 0;
 
-    while (pkts < UDX_SENDMMSG_BATCH_SIZE && socket->send_queue.len > 0) {
-      udx_packet_t *pkt = udx__fifo_shift(&(socket->send_queue));
-      /* pkt is null when descheduled after being acked */
-      if (pkt == NULL) {
-        if (pkts == 0) {
-          continue;
-        }
-        // return null to queue and send partial batch
-        // eliminates edge case where sendmmsg does
-        // a partial send and we must determine
-        // how many times to call udx__fifo_undo
-        udx__fifo_undo(&socket->send_queue);
-        break;
-      }
+    udx_packet_t *pkt = udx__fifo_peek(fifo);
+    int ttl = pkt->ttl;
+    bool adjust_ttl = ttl > 0 && socket->ttl != ttl;
+
+    while (pkts < UDX_SENDMMSG_BATCH_SIZE && fifo->len > 0) {
+      udx_packet_t *pkt = udx__fifo_peek(fifo);
+      // packet is null when descheduled after being acked
+      if (pkt == NULL || pkt->ttl != ttl) break;
+
+      udx__fifo_shift(fifo);
```

Comment:
I think so.

1. peek next packet in queue
2. if it's null or ttl is different then current batch, leave it on queue and send current batch
3. otherwise, add packet to batch and advance

the old code does shift instead of peek, but then does undo in the null case.

Reply:
ah yes i misread this as undo i think